sos/report/__init__.py

```diff
@@ -522,6 +524,131 @@ def _get_namespaces(self):
+        :rtype:     ``dict`` with keys being device types
+        """
+        _devs = {
+            'ethernet': [],
+            'bridge': [],
+            'team': [],
+            'bond': []
+        }
+        try:
+            if is_executable('nmcli', sysroot=self.opts.sysroot):
+                _devs.update(self._get_nmcli_devs())
+            # if nmcli failed for some reason, fallback
+            if not _devs['ethernet']:
+                self.soslog.debug(
+                    'Network devices not enumerated by nmcli. Will attempt to '
+                    'manually compile list of devices.'
+                )
+                _devs.update(self._get_eth_devs())
+                _devs['bridge'] = self._get_bridge_devs()
+        except Exception as err:
+            self.soslog.warn("Could not enumerate network devices: %s" % err)
+        return _devs
+
+    def _get_network_namespace_devices(self):
+        """Enumerate the network devices that exist within each network
+        namespace that exists on the system
+        """
+        _nmdevs = {}
+        for nmsp in self.namespaces['network']:
+            _nmdevs[nmsp] = {
+                'ethernet': self._get_eth_devs(nmsp)
+            }
+        return _nmdevs
+
+    def _get_nmcli_devs(self):
+        """Use nmcli, if available, to enumerate network devices. From this
+        output, manually grok together lists of devices.
+        """
+        _devs = {}
+        try:
+            _ndevs = sos_get_command_output('nmcli --fields DEVICE,TYPE dev')
+            if _ndevs['status'] == 0:
+                for dev in _ndevs['output'].splitlines()[1:]:
+                    dname, dtype = dev.split()
+                    if dtype not in _devs:
+                        _devs[dtype] = [dname]
+                    else:
+                        _devs[dtype].append(dname)
+                    _devs['ethernet'].append(dname)
+            _devs['ethernet'] = list(set(_devs['ethernet']))
+        except Exception as err:
+            self.soslog.debug("Could not parse nmcli devices: %s" % err)
+        return _devs
+
+    def _get_eth_devs(self, namespace=None):
+        """Enumerate a list of ethernet network devices so that plugins can
+        reliably iterate over the same set of devices without doing piecemeal
+        discovery.
+
+        These devices are used by `add_device_cmd()` when `devices` includes
+        "ethernet" or "network".
+
+        :param namespace:   Inspect this existing network namespace, if
+                            provided
+        :type namespace:    ``str``
+
+        :returns:   All valid ethernet devices found, potentially within a
+                    given namespace
+        :rtype:     ``list``
+        """
+        _eth_devs = []
+        if not namespace:
+            _eth_devs = [
+                dev for dev in listdir('/sys/class/net', self.opts.sysroot)
+                if dev not in ['bonding_masters']
```

Comment:
Isn't `if dev != 'bonding_master'` easier comparison?

Reply:
Initial thinking was that we'd have a handful of skip patterns, or at least one that might build up over time.